pkg/network/tracer/perf_batching.go

```diff
@@ -25,86 +31,135 @@ type PerfBatchManager struct {
 
 	// stateByCPU contains the state of each batch.
 	// The slice is indexed by the CPU core number.
-	stateByCPU []batchState
-
-	// maxIdleInterval represents the maximum time (in nanoseconds)
-	// a batch can remain idle (that is, without being flushed) on eBPF
-	maxIdleInterval int64
+	stateByCPU []percpuState
 }
 
 // NewPerfBatchManager returns a new `PerfBatchManager` and initializes the
 // eBPF map that holds the tcp_close batch objects.
-func NewPerfBatchManager(batchMap *ebpf.Map, maxIdleInterval time.Duration, numBatches int) (*PerfBatchManager, error) {
+func NewPerfBatchManager(batchMap *ebpf.Map, numCPUs int) (*PerfBatchManager, error) {
 	if batchMap == nil {
 		return nil, fmt.Errorf("batchMap is nil")
 	}
 
-	for i := 0; i < numBatches; i++ {
+	state := make([]percpuState, numCPUs)
+	for cpu := 0; cpu < numCPUs; cpu++ {
 		b := new(batch)
-		b.cpu = _Ctype_ushort(i)
-		batchMap.Put(unsafe.Pointer(&i), unsafe.Pointer(b))
+		batchMap.Put(unsafe.Pointer(&cpu), unsafe.Pointer(b))
+		state[cpu] = percpuState{
+			processed: make(map[uint64]batchState),
+		}
 	}
 
 	return &PerfBatchManager{
-		batchMap:        batchMap,
-		stateByCPU:      make([]batchState, numBatches),
-		maxIdleInterval: maxIdleInterval.Nanoseconds(),
+		batchMap:   batchMap,
+		stateByCPU: state,
 	}, nil
 }
 
 // Extract from the given batch all connections that haven't been processed yet.
 // This method is also responsible for keeping track of each CPU core batch state.
-func (p *PerfBatchManager) Extract(b *batch, now time.Time) []network.ConnectionStats {
-	if int(b.cpu) >= len(p.stateByCPU) {
+func (p *PerfBatchManager) Extract(b *batch, cpu int) []network.ConnectionStats {
+	if cpu >= len(p.stateByCPU) {
 		return nil
 	}
 
-	state := &p.stateByCPU[b.cpu]
-	lastOffset := state.offset
-	state.updated = now.UnixNano()
-	state.offset = 0
+	batchId := uint64(b.id)
+	cpuState := &p.stateByCPU[cpu]
+	start := uint16(0)
+	if bState, ok := cpuState.processed[batchId]; ok {
+		start = bState.offset
+	}
 
-	buffer := make([]network.ConnectionStats, 0, ConnCloseBatchSize)
-	return ExtractBatchInto(buffer, b, lastOffset, ConnCloseBatchSize)
+	buffer := make([]network.ConnectionStats, 0, ConnCloseBatchSize-start)
+	conns := p.extractBatchInto(buffer, b, start, ConnCloseBatchSize)
+	if len(conns) > 0 {
+		cpuState.processed[batchId] = batchState{offset: ConnCloseBatchSize, updated: time.Now()}
+	}
+
+	p.cleanupExpiredState()
```

Comment:
This is in a hot code path and I think there is no good reason to keep it here. How about just leaving it in `GetIdleConns()` which gets periodically called?

Reply:
On a second thought the map could get quite large in a server with high connection churn. So perhaps I'd consider just throttling this.